node/network/availability-distribution/src/tests.rs

```diff
@@ -1081,3 +1051,222 @@ fn k_ancestors_in_session() {
+
+	state.receipts.insert(hash_a, HashSet::new());
+	state.receipts.insert(hash_b, HashSet::new());
+	state.receipts.insert(hash_c, HashSet::new());
+	state.receipts.insert(hash_d, HashSet::new());
+
+	state.per_relay_parent.insert(hash_a, PerRelayParent {
+		ancestors: vec![hash_b],
+		live_candidates: HashSet::new(),
+	});
+
+	state.per_relay_parent.insert(hash_c, PerRelayParent::default());
+
+	state.clean_up_receipts_cache();
+
+	assert_eq!(state.receipts.len(), 3);
+	assert!(state.receipts.contains_key(&hash_a));
+	assert!(state.receipts.contains_key(&hash_b));
+	assert!(state.receipts.contains_key(&hash_c));
+	assert!(!state.receipts.contains_key(&hash_d));
+}
+
+#[test]
+fn remove_relay_parent_only_removes_per_candidate_if_final() {
+	let mut state = ProtocolState::default();
+
+	let hash_a = [0u8; 32].into();
+	let hash_b = [1u8; 32].into();
+
+	let candidate_hash_a = CandidateHash([46u8; 32].into());
+
+	state.per_relay_parent.insert(hash_a, PerRelayParent {
+		ancestors: vec![],
+		live_candidates: std::iter::once(candidate_hash_a).collect(),
+	});
+
+	state.per_relay_parent.insert(hash_b, PerRelayParent {
+		ancestors: vec![],
+		live_candidates: std::iter::once(candidate_hash_a).collect(),
+	});
+
+	state.per_candidate.insert(candidate_hash_a, PerCandidate {
+		live_in: vec![hash_a, hash_b].into_iter().collect(),
+		..Default::default()
+	});
+
+	state.remove_relay_parent(&hash_a);
+
+	assert!(!state.per_relay_parent.contains_key(&hash_a));
+	assert!(!state.per_candidate.get(&candidate_hash_a).unwrap().live_in.contains(&hash_a));
+	assert!(state.per_candidate.get(&candidate_hash_a).unwrap().live_in.contains(&hash_b));
+
+	state.remove_relay_parent(&hash_b);
+
+	assert!(!state.per_relay_parent.contains_key(&hash_b));
+	assert!(!state.per_candidate.contains_key(&candidate_hash_a));
+}
+
+#[test]
+fn add_relay_parent_includes_all_live_candidates() {
+	let relay_parent = [0u8; 32].into();
+
+	let mut state = ProtocolState::default();
+
+	let ancestor_a = [1u8; 32].into();
+
+	let candidate_hash_a = CandidateHash([10u8; 32].into());
+	let candidate_hash_b = CandidateHash([11u8; 32].into());
+
+	let candidates = vec![
+		(candidate_hash_a, FetchedLiveCandidate::Fresh(Default::default())),
+		(candidate_hash_b, FetchedLiveCandidate::Cached),
+	].into_iter().collect();
+
+	state.add_relay_parent(
+		relay_parent,
+		Vec::new(),
+		None,
+		candidates,
+		vec![ancestor_a],
+	);
+
+	assert!(
+		state.per_candidate.get(&candidate_hash_a).unwrap().live_in.contains(&relay_parent)
+	);
+	assert!(
+		state.per_candidate.get(&candidate_hash_b).unwrap().live_in.contains(&relay_parent)
+	);
+
+	let per_relay_parent = state.per_relay_parent.get(&relay_parent).unwrap();
+
+	assert!(per_relay_parent.live_candidates.contains(&candidate_hash_a));
+	assert!(per_relay_parent.live_candidates.contains(&candidate_hash_b));
+}
+
+#[test]
+fn query_pending_availability_at_pulls_from_and_updates_receipts() {
+	let hash_a = [0u8; 32].into();
```

Comment:
There is `Hash::repeat_byte` which is used elsewhere.

Reply:
`Hash::repeat_byte(0)` is longer than
`[0u8; 32].into()`

FWIW

Reply:
But invariant over the hash length :)